### PR TITLE
Block BlobStore from starting if it was down for X days.

### PR DESCRIPTION
**Pull request by JingQianCloud**
Block BlobStore from starting if it was down for X days.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java [hunk 1]
```diff
@@ -650,6 +650,20 @@ public class StoreConfig {
...
+  @Config(storeStaleTimeInDaysName)
+  public final int storeStaleTimeInDays;
+  public final static String storeStaleTimeInDaysName = "store.stale.time.in.days";
+
+  /**
+   * If the store is stale and storeStaleBlockBootup is true, don't start the blob.
+   */
+  @Config(storeStaleBlockBootupName)
+  public final boolean storeStaleBlockBootup;
+  public final static String storeStaleBlockBootupName = "store.stale.block.bootup";
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java:665]: let's rename this variable to storeBlockStaleBlobStoreToStart.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 2]
```diff
@@ -324,6 +329,52 @@ public void start() throws StoreException {
...
+    String lastAliveTime = formatter.format(date);
+    logger.info(
+        "checkIfStoreIsStale " + replicaId.getReplicaPath() + " last modified file " + ((lastModifiedFile != null)
+            ? lastModifiedFile.getAbsolutePath() : "") + " last alive time " + lastAliveTime);
+
+    if (config.storeStaleTimeInDays != 0
+        && lastModifiedTimeInMilliseconds + TimeUnit.DAYS.toMillis(config.storeStaleTimeInDays) < time.milliseconds()) {
+      logger.error("checkIfStoreIsStale " + replicaId.getReplicaPath() + " is stale. ");
+      if (config.storeStaleBlockBootup) {
+        throw new StoreException("BlobStore is stale ", StoreErrorCodes.Initialization_Error);
```
Comment (justinlin-linkedin) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:373]: Let's add the dataDir to the message of this exception.

Comment (justinlin-linkedin) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:373]: Also, can we add a GAUGE metric to show the number of stale blobstores.

Comment (JingQianCloud) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java]: Added the Gauge and addressed all the comments. thanks,

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-store/src/main/java/com/github/ambry/store/StoreMetrics.java [hunk 3]
```diff
@@ -232,6 +232,8 @@ public StoreMetrics(String prefix, MetricRegistry registry) {
     duplicateKeysInBatch = registry.counter(MetricRegistry.name(BlobStore.class, name + "DuplicateKeysInBatch"));
     storeIoErrorTriggeredShutdownCount =
         registry.counter(MetricRegistry.name(BlobStore.class, name + "StoreIoErrorTriggeredShutdownCount"));
+    Gauge<Integer> staleBlobStoreCount = BlobStore.staleBlobCount::get;
+    registry.register(MetricRegistry.name(BlobStore.class, name + "StaleBlobStoreCount"), staleBlobStoreCount);
```
Comment (justinlin-linkedin) [inline ambry-store/src/main/java/com/github/ambry/store/StoreMetrics.java:236]: let's do 
```
registry.gauge(MetricRegistry.name(BlobStore.class, name + "StaleBlobStoreCount"), () -> staleBlobStoreCount);
```

Comment (JingQianCloud) [reply ambry-store/src/main/java/com/github/ambry/store/StoreMetrics.java]: It's a magic. :-) 
Thanks,

**Pull request merged**